### PR TITLE
[TypeSpec Validation] Add mechanism for specs to temporarily opt-out of "all spec" validation

### LINKED ISSUE #24955: [TypeSpec Validation] Add per-rule suppression mechanism
Add ability for specs to opt-out of TSV rules, using the common `specifications.yaml` mechanism we are planning.

Depends on: https://github.com/Azure/azure-rest-api-specs/issues/28447

### PR DESCRIPTION
**Pull request by ckairen**
closes #24955

### CODE REVIEW

File: eng/scripts/Get-TypeSpec-Folders.ps1 [hunk 1]
```diff
@@ -42,7 +42,8 @@ else {
 $typespecFolders = @()
 foreach ($file in $changedFiles) {
   if ($file -match 'specification\/[^\/]*\/') {
-    $typespecFolder = (Get-ChildItem -path $matches[0] tspconfig.yaml -Recurse).Directory.FullName | ForEach-Object {if ($_) { [IO.Path]::GetRelativePath($($pwd.path), $_) }}
+    $typespecFolder = (Get-ChildItem -path $matches[0] tspconfig.yaml -Recurse).Directory.FullName
+      | ForEach-Object {if ($_ -and !(Get-ChildITem -force -path $_ .skip-typespec-validation)) { [IO.Path]::GetRelativePath($($pwd.path), $_) }}
```
Comment (mikeharder) [inline eng/scripts/Get-TypeSpec-Folders.ps1:46]: Move filtering to only apply to `$allChangedFiles`

**Review by mikeharder — CHANGES_REQUESTED** [review]

File: eng/scripts/Get-TypeSpec-Folders.ps1 (hunk 1 above)
Comment (ckairen) [reply eng/scripts/Get-TypeSpec-Folders.ps1]: this won't give us the correct result. 
The code on line 45
```
if ($file -match 'specification\/[^\/]*\/') {
    $typespecFolder = (Get-ChildItem -path $matches[0] tspconfig.yaml -Recurse).Directory.FullName
    | ForEach-Object {if ($_) { [IO.Path]::GetRelativePath($($pwd.path), $_) }}
```
will add the folder in again. (because of the logic for "changed file should trigger all folders beyond depth 2 folders"
e.g. changedFiles of specification/cognitiveservices/AnomalyDetector should trigger all folders under cognitiveservices.)

In addition, we also want to separate the $allChangedFiles into two categories, one is the ci run, another is the pr change that has an eng folder or root folder change. there's an edge case of a change in both eng folder and skipped test spec directory

File: eng/scripts/Get-TypeSpec-Folders.ps1 [hunk 2]
```diff
@@ -12,7 +12,8 @@ $changedFiles = @()
 $allChangedFiles = (Get-ChildItem -path ./specification tspconfig.yaml -Recurse).Directory.FullName | ForEach-Object {[IO.Path]::GetRelativePath($($pwd.path), $_)}
 $allChangedFiles = $allChangedFiles -replace '\\', '/'
 
-if ([string]::IsNullOrEmpty($TargetBranch) -or [string]::IsNullOrEmpty($SourceBranch)) {
+$ci = [string]::IsNullOrEmpty($TargetBranch) -or [string]::IsNullOrEmpty($SourceBranch)
```
Comment (weshaggard) [inline eng/scripts/Get-TypeSpec-Folders.ps1:15]: What if someone runs this locally with an invalid target or source? It wouldn't be a CI run. 

The common way we detect running in a DevOps job is by looking this env variable https://github.com/Azure/azure-sdk-tools/blob/main/eng/common/scripts/logging.ps1#L3

File: eng/scripts/Get-TypeSpec-Folders.ps1 [hunk 3]
```diff
@@ -42,7 +43,12 @@ else {
 $typespecFolders = @()
 foreach ($file in $changedFiles) {
   if ($file -match 'specification\/[^\/]*\/') {
-    $typespecFolder = (Get-ChildItem -path $matches[0] tspconfig.yaml -Recurse).Directory.FullName | ForEach-Object {if ($_) { [IO.Path]::GetRelativePath($($pwd.path), $_) }}
+    $typespecFolder = (Get-ChildItem -path $matches[0] tspconfig.yaml -Recurse).Directory.FullName
+    | ForEach-Object {if ($_) { [IO.Path]::GetRelativePath($($pwd.path), $_) }}
+
+    if ($ci) {
+      $typespecFolder = $typespecFolder | Where-Object {!(Get-ChildITem -force -path $_ .skip-typespec-validation)}
```
Comment (weshaggard) [inline eng/scripts/Get-TypeSpec-Folders.ps1:50]: Is the expectation that people put this empty marker file next to their tspconfig?

Comment (ckairen) [reply eng/scripts/Get-TypeSpec-Folders.ps1:50]: yes that's the design so far. We thought about putting it in the tspconfig but it doesn't quite fit into the definition of "configuring tsp projects"

File: eng/scripts/Get-TypeSpec-Folders.ps1 (hunk 2 above)
Comment (ckairen) [reply eng/scripts/Get-TypeSpec-Folders.ps1:15]: we also want locally runs of "run all" to skip these files as well. but a "runall" variable might be confusing between ci/runall option and runall due to eng and root dir

**Comment by mikeharder** [comment]
@ckairen: I am closing this PR and moving this issue to the backlog, for a few reasons:

1. I'm not sure if we will need this feature, due to my fix to #25175.
2. Even if we do need this feature for other reasons, I don't think we'd want to disable validation for so many specs in the private repo.
3. If we do implement this feature, I think we'll want to use a common config file named `tsvconfig.yaml` (as a sibling to `tspconfig.yaml`), with a property like `enabled: false`.
